KAFKA-7739: Tiered storage

satishd opened the pull request:
[WIP] This is the initial **draft** version of the KIP-405. It includes the initial set of changes required for plugging in a `RemoteStorageManager`. We will update KIP-405 and this PR in the next few days with more details.

KIP-405 is located at https://s.apache.org/pk53b

This PR contains HDFS/S3  implementations to discuss `RemoteStorageManager` APIs and it will also be easier to review and test the proposed APIs end to end. This implementation will be removed before it is ready to be merged. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

satishd committed:
KAFKA-7739: Added basic remote log manager related traits/classes and bootstrapping.

satishd committed:
Added basic logcopying in remote log manager using RemoteLogStorageManager

satishd committed:
More updates with API and usage

satishd committed:
Added remote log/offset/timestamp indexes

satishd committed:
Updated with minor fixes

satishd committed:
Added parsing of entries from remote log index file

satishd committed:
Minor fixes for config types and respective tests

satishd committed:
Fixed test to check for possible exception

ying-zheng committed:
add .arcconfig file to tiered storage branch

ying-zheng committed:
Add RSM module

ying-zheng committed:
RSM interface changes

Reviewers: satishd, harshach

Subscribers: #streaming_data

Differential Revision: https://code.uberinternal.com/D2894829

ying-zheng committed:
Allow creating RemoteLogIndexEntry without crc

Summary:
1. Add a RemoteLogIndexEntry "constructor", to allow calculating crc inside the RemoteLogIndexEntry class
2. Index entry length should be 16 bits
3. Allow RSM read the content of RDI

Reviewers: harshach, satishd

Reviewed By: harshach, satishd

Subscribers: jenkins, #streaming_data

Differential Revision: https://code.uberinternal.com/D2911255

satishd committed:
Added periodic task to poll topic-partition log directories to add passive segments for copying

Summary: Added periodic task to poll topic-partition log directories to add passive segments for copying

Reviewers: harshach, yingz

Reviewed By: harshach

Differential Revision: https://code.uberinternal.com/D2921725

ying-zheng committed:
Implement HDFS RSM

Summary:
1. Implement HDFS RSM
2. Some changes to the RSM interface:
    2.1 throw IOException in most methods
    2.2 copyLogSegment do not need to return RDI, cause its already in RemoteIndexEntry
    2.3 read method need RemoteIndexEntry in addition to RDI
3. Move RemoteLogIndex.parseEntry to RemoteLogIndexEntry.parseEntry. This method is also needed by HDFS RSM

Reviewers: harshach, satishd

Reviewed By: harshach

Subscribers: jenkins, #streaming_data

Differential Revision: https://code.uberinternal.com/D2960007

satishd committed:
Added initial implementation of RLM follower

Summary:
Added initial implementation of RLM follower
- Refactored the existing indexes for RLM follower
- Followup PRs will address remaining issues about boundary cases of adding indexes for conflicting offsets  and scale up with fine-grained locking

Reviewers: harshach, yingz

Reviewed By: yingz

Subscribers: jenkins

Differential Revision: https://code.uberinternal.com/D2972601

satishd committed:
T3305897: Initial version of fetch implementation for RemoteLogManager. It includes inter broker fetch protol changes for sending local log offset when the requested offset is in remote tier. This will allow follower broker to know that the...

Summary:
Initial version of fetch implementation for RemoteLogManager.
- It includes inter broker fetch protol changes for sending local log offset when the requested offset is in remote tier. This will allow follower broker to know that therequested offset is moved to remote tier and it can start fetching the messages available in local log from the leader.
- Added test for RLM fetch API.
- Fixed spotbug errors with recent additions.

Reviewers: harshach, yingz

Reviewed By: harshach, yingz

Differential Revision: https://code.uberinternal.com/D3014447

ying-zheng committed:
Copy dependent jars when building hdfs rsm

Summary:
1. Copy dependent jars in the same way as Kafka core and the other modules do.
2. Remove the dependency on slf4japi, which is already included in Kafka core dependency

Reviewers: satishd, harshach

Reviewed By: satishd, harshach

Subscribers: #streaming_data_kafka

Differential Revision: https://code.uberinternal.com/D3035243

satishd committed:
Local segments are scheduled to be deleted once they are copied to remote tier storage
- Fixed issues in finding the right local segments while copying to remote tier storage

Summary:
Local segments are scheduled to be deleted once they are copied to remote tier storage.
  - Fixed issues in finding the right local segments while copying to remote tier storage .

Reviewers: harshach, yingz

Reviewed By: harshach

Subscribers: jenkins

Differential Revision: https://code.uberinternal.com/D3046097

ying-zheng committed:
Use java list instead of Seq in RSM interface

Summary:
1. Use java list instead of Seq in RSM interface
2. Sort remote log segments in RemoteStorageManager.listRemoteSegments

Reviewers: satishd, harshach

Reviewed By: satishd

Subscribers: jenkins, #streaming_data_kafka

Differential Revision: https://code.uberinternal.com/D3102715

ying-zheng committed:
Format HDFSRemoteStorageManagerTest.java

Summary: Fix some format issues reported by CheckStyle

Reviewers: satishd, harshach

Reviewed By: harshach

Subscribers: #streaming_data_kafka

Differential Revision: https://code.uberinternal.com/D3160621

ivanyu committed:
Add ability list remote segments from min offset

This commit adds a new variant of listRemoteSegments method to RemoteStorageManager, which has minBaseOffset parameter. This is useful for reducing listing time on the remote tier in some implementations (like S3).

ying-zheng committed:
Add remote log reader thread pool

ying-zheng committed:
Add new config to KafkaConfigTest

ying-zheng committed:
Add RemoteLogReaderTest, fix cancel remote read task

ying-zheng committed:
Add apache license RemoteLogReaderTest.scala

ying-zheng committed:
Add remote log reader thread pool and purgatory

Summary: Handle remote storage fetch requests in a separated thread pool, to avoid blocking Kafka I/O threads. Added a new purgatory to handle timeout.

Reviewers: satishd, harshach

Subscribers: #streaming_data_kafka

Differential Revision: https://code.uberinternal.com/D3224503

satishd committed:
T3788999: RLM enhancements for copying segments, syncing indexes and cleaningup expired remote log segments based on retention period by using tasks run at regular intervals.

Summary:
RLM enhancements for copying segments, syncing indexes and cleaningup expired remote log segments based on retention period by using tasks run at regular intervals.
 Added configs for task thread pool size and interval.

 Need to add remote log startoffset handling, which will be done in a followup PR/diff.

Reviewers: harshach, yingz

Reviewed By: harshach, yingz

Subscribers: jenkins

Differential Revision: https://code.uberinternal.com/D3211435

satishd committed:
T3064289: Handled Followers to cleanup expired remote log indexes.

Summary:
T3064289 Handled followers to cleanup expired remote log indexes.
- Added a method to get earliest log offset in RSM
- Refactored TopicPartitionIndex.

Reviewers: harshach

Reviewed By: harshach

Subscribers: jenkins

Differential Revision: https://code.uberinternal.com/D3270551

ying-zheng committed:
Validate the new configuration

ying-zheng committed:
Fix unit test package name

ivanyu committed:
Allow to read RemoteLogIndexEntry from InputStream

This commit adds new methods to `RemoteLogIndexEntry` that allow entries to be read from `InputStream`, not only from `FileChannel`. This helps to avoid writing temporary files in `HDFSRemoteStorageManager` and other implementations: now it's possible to parse remote log index entries from the network input stream directly.
Some tests were added to test this functionality along with the previous (reading from `FileChannel`).

ying-zheng committed:
Added comments to ReplicaManager.readFromLocalLog()

Differential Revision: https://code.uberinternal.com/D3242859

satishd committed:
T3839003: Added localLogStartOffset, logStartOffset for both local and effective LSO of a topic partition even when remote storage is enabled.

Summary:
T3839003:  Added localLogStartOffset, logStartOffset for both local and effective LSO of a topic partition even when remote storage is enabled.
RLM copies segments which are earlier to min of highWatermark and logStableOffset.
Added the removal of remote log indexes when rmeote log-start-offset updates are received.
Refactored RemoteLogManager creating in KafkaServer and LogManager's dependency on RLM.

Reviewers: harshach, yingz

Subscribers: jenkins

Differential Revision: https://code.uberinternal.com/D3245381

ying-zheng committed:
Implement HDFS RSM deleteTopicPartition() and cleanupLogUntil()

Reviewers: satishd, harshach

Subscribers: jenkins, #streaming_data_kafka

Differential Revision: https://code.uberinternal.com/D3274317

satishd committed:
T3880979 Refactor RemoteLogSegmentInfo to be an immutable object and it should contain all the information required to know about a segment.

Summary: T3880979 Refactor RemoteLogSegmentInfo to be an immutable object and it should contain all the information required to know about a segment.

Reviewers: harshach

Reviewed By: harshach

Subscribers: jenkins

Differential Revision: https://code.uberinternal.com/D3281345

satishd committed:
T3898121 Pass local partition leader epoch value to RSM while invoking copyLogSegments.

Summary: T3898121 Pass local partition leader epoch value to RSM while invoking copyLogSegments.

Differential Revision: https://code.uberinternal.com/D3313275

satishd committed:
Renamed endOffset to lastOffset

satishd committed:
MINOR Pulled/rebased latest trunk changes and fixed merge conflicts.

Differential Revision: https://code.uberinternal.com/D3337121

satishd committed:
MINOR Fixed copying log segments of a topic partition in order.
 - Fixed HDFS RSM to return right offset values when there are no remote log segments for a topic partition.

satishd committed:
MINOR Fixed checkstyle issues in remote-storage-managers:hdfs module

satishd committed:
MINOR Fixed building remote log indexes when a partition remote index is reopened

satishd committed:
MINOR Fixed building indexes for the passive segments which are not yet read.
- Fixed FetchRequest and FetchResponse jsons with right field names and versions.
- Fixed updating log-start-offset when remote storage is enabled.
- Fixed TopicPartitionRemoteIndexTest

satishd committed:
T4223245 Delete local segments only if the remote log indexes exists locally till that offset.

satishd committed:
MINOR Avoid duplicate index closes by removing closing indexes before rename as these indexes are closed later in deleteIfExists

satishd committed:
MINOR Excluded .arcconfig from rat checks

satishd committed:
T4242585 Enable tiered storage only for non compacted topics.

satishd committed:
MINOR Invoke stopPartitions when a partition replica is stopped/deleted

satishd committed:
Merge remote-tracking branch 'apache-repo/trunk' into sato-tiered-apache-rebase

satishd committed:
MINOR Fixed minor build failures caused while merging upstream apache/trunk changes.

satishd committed:
MINOR Fixed LogCleanerIntegrationTest.testIsThreadFailed by removing the metric if it exists

satishd committed:
Merge remote-tracking branch 'apache-repo/trunk' into sato-tiered-apache-rebase

satishd committed:
MINOR Fixed a few files which come from upstream apache/trunk by updating them with the introduced API changes from this branch.

satishd committed:
MINOR Updated the javadocs of RemoteStorageManager and RemoteLogSegmentInfo.

satishd committed:
Merge remote-tracking branch 'apache-repo/trunk' into sato-tiered-apache-rebase

ambud commented on .arcconfig:
```diff
@@ -0,0 +1,5 @@
+{
```
Probably should be removed

satishd committed:
MINOR Removed cancelCopyLogSegments from RemoteStorageManager and addressed other minor review comments
 - Removed .arcconfig as it is nto relevant in apache/kafka repo.
 - Added initial README.md in remotre-storage-managers/hdfs.

ying-zheng committed:
RSM class loader

ying-zheng committed:
Address review comments of RSM class loader

enothereska commented on core/src/main/scala/kafka/log/remote/TopicPartitionRemoteIndex.scala:
```diff
@@ -0,0 +1,236 @@
+ *  - time index
+ *   - maintains a mapping between timestamp and position in the remote log index which contains remote log information for this timestamp.
+ *
+ *   These index files are also locally stored in respective log-dir of the topic partition.
```
Can these index files be reconstructed in case they are corrupted for whatever reason, e.g., unclean shutdown?

satishd replied:
Right, these can be reconstructed using the below APIs. This is a pending `todo` as mentioned [here](https://github.com/apache/kafka/pull/7561/files/7c9abf3a3bca2024092bd823f3225dafc8e84bb9#diff-eeb791e578ced303c3141c5369366f80R103)

```
listRemoteSegments(topicPartition: TopicPartition, minBaseOffset: Long)
getRemoteLogIndexEntries(remoteLogSegment: RemoteLogSegmentInfo): util.List[RemoteLogIndexEntry]
```

Hangleton replied:
Do you plan to hook some sort of recovery of remote indexes in `remote.log.LogSegment#recover`? However, since only inactive segments are on the remote storage, I wonder to which extent they should be involved in the recovery process.

ying-zheng committed:
Merge pull request #19 from ying-zheng/rsm_classloader

RSM class loader

ying-zheng committed:
Set context class loader to RSM class loader when calling RSM methods

ying-zheng committed:
The initial value of RLMTask.readOffset should be -1

ying-zheng committed:
Filter remote segments by endOffset rather than baseOffset (#22)

RemoteStorageManager.listRemoteSegments() filters remote segments by
endOffset instead of baseOffset

satishd committed:
MINOR Fixed spotbugs error.

satishd committed:
MINOR Few checkstyle fixes.

satishd committed:
MINOR Added missing license header for RemoteStorageManagerWrapper.

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -0,0 +1,467 @@
+              sortedSegments.slice(0, index).foreach { segment =>
+                // store locally here as this may get updated after the below if condition is computed as false.
+                val leaderEpochVal = leaderEpoch
+                if (isCancelled() || !isLeader) {
```
Is any processing in `copyLogSegmentsToRemote` required if the broker is not the leader for the topic-partition? Or could the condition `!isLeader` be moved line 225?

satishd replied:
`isLeader` check is done for each segment and if it is moved from leader to follower or it is removed from replicas then it stops copying the remaining segments and returns from here. If this instance has become a follower then `copyLogSegmentsToRemote` is not done in subsequent runs because of the below check in `run()` method.

```
 if (isLeader) copyLogSegmentsToRemote()
```

Hangleton replied:
I see, so this local check avoid copying segments if a reassignment has happened for the topic-partition while `copyLogSegmentsToRemote` is being executed.
Actually I am trying to understand the role of `leaderEpoch` here. Is it for potential de-dupe?

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -0,0 +1,467 @@
+  }
+}
+
+trait CancellableRunnable extends Runnable {
```
Out of curiosity, what is the motivation behind introducing a `CancellableRunnable`? Could the `RLMTask` not rely on the interruption flag of the running thread (set by `Future#cancel(true)`) to detect cancellation?

satishd replied:
`RLMTask` can not depend on the interrupted flag as that may get reset by `RemoteStorageManager` in any of the operations used in `RLMTask`.

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -0,0 +1,467 @@
+class RLMScheduledThreadPool(poolSize: Int) extends Logging {
+
+  private val scheduledThreadPool: ScheduledThreadPoolExecutor = {
+    val threadPool = new ScheduledThreadPoolExecutor(poolSize)
```
As in `RemoteStorageThreadPool`, it may be worth to add metrics on queue size and utilization here?
That said, the workload on this thread pool is different from `RemoteStorageThreadPool` - driven by fetch requests for the latter, and the number of topic-partitions for the former, and may thus not justify the same level of observability.

satishd replied:
We plan to add more metrics later like how much time is being taken for copying segments etc but not yet finalized the list to be added.

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteStorageManager.scala:
```diff
@@ -0,0 +1,138 @@
+ * All these APIs are still experimental.
+ */
+@InterfaceStability.Unstable
+trait RemoteStorageManager extends Configurable with AutoCloseable {
```
If I understand correctly, this interface defines a set of CRUD operations on the remote log segments. I would drop the "Manager" from the name. It is true that it is not really a client or proxy to the remote storage system, as it embeds a precisely defined application semantic, though.

satishd replied:
I prefer to keep this class name as `RemoteLogStorageManager`. I would like to hear others opinions too.

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteStorageManager.scala:
```diff
@@ -0,0 +1,138 @@
+ * All these APIs are still experimental.
+ */
+@InterfaceStability.Unstable
+trait RemoteStorageManager extends Configurable with AutoCloseable {
```
I wonder if implementations of `RemoteStorageManager` have to guarantee strong consistency for read-after-write. For instance with S3, there are some corner cases where the storage is eventually consistent [[1]](https://docs.aws.amazon.com/AmazonS3/latest/dev/Introduction.html#ConsistencyModel), but from a first look at the S3 PR (review in progress) I did not see any such cases which could be exercised.

But, eventually consistent storage may not cause any problem anyway. I did not see a case in the `RLMTask` when the update of remote log indexes [[2]](https://github.com/apache/kafka/pull/7561/files#diff-a597bd0c7d627789e73d1fa38eb1abfaR282) could be inconsistent with the indexes returned from the copy [[3]](https://github.com/apache/kafka/pull/7561/files#diff-a597bd0c7d627789e73d1fa38eb1abfaR258) because there should be not overlap in the segments uploaded and listed thereafter since the `readOffset` was updated to the last remote log offset.

One case where eventual consistency could intervene is if a read from a Fetch request happens on a segment just uploaded to the remote storage; but in such cases, the read would fail but will eventually succeed after retries (with enough backoff), which should be fine.

ivanyu replied:
@Hangleton In case you're interested, there's a [proposed `S3RemoteStorageManager` implementation](https://github.com/harshach/kafka/pull/18/files) that's being reviewed. Currently, it doesn't do file overwrites on S3, so the only state that is eventually being reached is file existence. This can be solved with some sensible retry policy.

We'll have to invent some consistency mechanism if overwriting starts happening at some point.

*UPD:* sorry, I've noticed later you've seen it already.

satishd replied:
@Hangleton @ivanyu RLM expects remote logs to be available once they are copied successfully. But it looks like S3 gives eventual consistency model which may cause reads to fail as that object may not yet be available. Are there any guarantees that the uploaded object will be available after a certain duration? As the committed log segments are copied as soon as they are available, this issue may not arise in normal scenarios but it is better to be discussed and set the right contracts with the API.  

There can be other issues like create a topic again after a delete but earlier remote log segments in S3 may still be available which can cause issues. [KIP-516](https://cwiki.apache.org/confluence/display/KAFKA/KIP-516%3A+Topic+Identifiers) gives topic-id as identifier which resolves uniqueness of a topic. This feature may be helpful for create after delete scenario but it is not yet available in Kafka.

Did you think of addressing these kind of issues by maintaining remote log metadata in a key/value store with strong consistency or do you want to mitigate these kind of issues with a retry policy?

ivanyu replied:
> Are there any guarantees that the uploaded object will be available after a certain duration?

Unfortunately, it's unbounded.

Hangleton replied:
Yes the time for convergence is indeterminate and we cannot rely on retries for that.
The consistency model for S3 is

- Strongly consistent on PUT only if no HEAD or GET requests were issued on the key name before creating the object.
- Eventually consistent for PUT overwriting an existing object.
- Eventually consistent for LIST after a PUT.
- Eventually consistent for LIST and GET after a DELETE object.

In this case if the `RemoteStorageManager` requires strong consistency after a mutating call, an additional index would need to be used.

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -0,0 +1,467 @@
+      }
+    }
+
+    override def run(): Unit = {
```
I wonder here what would the cost in traffic of each `RLMTask` be if the period of execution is "small" (what would be the order of magnitude of a typical value for `remote.log.manager.task.interval.ms`?). For instance, from a quick look at the S3 implementation (but I likely missed something at this stage), we would have a minimum of two requests to S3 to list objects through execution of `updateRemoteLogIndexes()` and `handleExpiredRemoteLogSegments()`. If we have, say 1k topic-partitions and 1 minute for `remote.log.manager.task.interval.ms`, we would have ~30 queries/s, which in case of S3 is high.

I guess it is the responsibility of implementations of `RemoteStorageManager` to provide efficient read for methods such as [`listRemoteSegments(.)`](https://github.com/apache/kafka/pull/7561/files#diff-526659056b3b2cd6ab1fe9db1977b826R79) and [`cleanupLogUntil `](https://github.com/apache/kafka/pull/7561/files#diff-526659056b3b2cd6ab1fe9db1977b826R119) (at the cost of strong read-after-write consistency) to limit traffic to the remote storage.

ivanyu replied:
Thank you for pointing this out. Maybe some form of batching requests across topic-partitions will be needed.

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteLogManager.scala:
```diff
@@ -0,0 +1,467 @@
+  private val remoteStorageFetcherThreadPool = new RemoteStorageReaderThreadPool(rlmConfig.remoteLogReaderThreads,
+    rlmConfig.remoteLogReaderMaxPendingTasks, time)
+
+  private val delayInMs = rlmConfig.remoteLogManagerTaskIntervalMs
```
cf. comment on `RLMTask` - what would be a typical delay?

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteLogIndex.scala:
```diff
@@ -0,0 +1,186 @@
+
+  /**
+   */
+  def reset(): Unit = {
```
Is this method used? Offset, time and transaction indexes use it when a segment is recovered and indexes are rebuilt, but remote indexes don't seem to in-place recovery? Or is this going to be part of the recovery mentioned [here](https://github.com/apache/kafka/pull/7561/files/7c9abf3a3bca2024092bd823f3225dafc8e84bb9#diff-eeb791e578ced303c3141c5369366f80R103)?

satishd replied:
RIght, recovery of remote indexes is a todo item as mentioned in the code. We plan to address this soon.

Hangleton replied:
ok, I guess this will contribute to address Jun's point 43.4 about recovery for JBOD.

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteLogIndex.scala:
```diff
@@ -0,0 +1,186 @@
+  }
+
+  def lastOffset: Option[Long] = {
+    _lastOffset
```
I wanted to ask why access to `_lastOffset` is unguarded but saw it is the same in `kafka.log.OffsetIndex`. I don't understand how memory visibility is guaranteed outside of `lock` and without `@volatile`, do you know?

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteLogIndexEntry.scala:
```diff
@@ -0,0 +1,229 @@
+ * @param dataLength     length of the data stored in remote tier at rdi.
+ * @param rdi            bytes value of rdi.
+ */
+case class RemoteLogIndexEntry(magic: Short, crc: Int, firstOffset: Long, lastOffset: Long,
```
Out of curiosity, what would happen in case of corrupted `RemoteLogIndexEntry` with an invalid CRC code? Here it seems it is used as part of the object's identity - do you plan to perform explicit check to detect potential inconsistencies?

Hangleton reviewed (commented):
Few comments, review still in progress.

Hangleton commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -313,10 +325,15 @@ class Log(@volatile var dir: File,
       s"log end offset $logEndOffset in ${time.milliseconds() - startMs} ms")
   }
 
+  def updateLogStartOffsetFromRemoteTier(remoteLso: Long): Unit = {
```
Maybe use another acronym here as LSO is already used for the Last Stable Offset?

satishd committed:
MINOR Addressed a minor comment and other trivial changes.

satishd committed:
MINOR Addressed a minor comment.

ying-zheng committed:
Bug fix: ship segment with only one record to remote storage

ying-zheng committed:
Merge pull request #23 from ying-zheng/single_record_segment

Bug fix: ship segment with only one record to remote storage

ying-zheng committed:
Extend list offset API to return next local offset

- Remove next local offset from fetch response.
- Extend list offset API to return next local offset.
- Fix message json files for next loal offset.

ying-zheng committed:
Change some log message levels from info to debug in RSMTask (#25)

ying-zheng committed:
Offset / Timestamp lookup for remote data

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteLogIndexEntry.scala:
```diff
@@ -0,0 +1,229 @@
+  }
+
+  def asBuffer: ByteBuffer = {
+    val buffer = ByteBuffer.allocate(entryLength + 2 + 2)
```
Maybe use `totalLength`?

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteLogIndexEntry.scala:
```diff
@@ -0,0 +1,229 @@
+      val entryOption = parseEntry(ch, pos.toInt)
+      val entry = entryOption.getOrElse(throw new KafkaException(s"Entry could not be found at position: $pos"))
+      index = index :+ entry
+      pos += entry.entryLength + 4
```
Maybe use `totalLength`?

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteLogIndexEntry.scala:
```diff
@@ -0,0 +1,229 @@
+    var pos = 0L;
+    while (pos < ch.size()) {
+      val entryOption = parseEntry(ch, pos.toInt)
+      val entry = entryOption.getOrElse(throw new KafkaException(s"Entry could not be found at position: $pos"))
```
Kafka uses `CorruptIndexException` for sanity check of offset indexes. Maybe something following the same pattern here e.g. `CorruptRemoteIndexException`?

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteLogIndexEntry.scala:
```diff
@@ -0,0 +1,229 @@
+          }
+        case _ =>
+          // TODO: Throw Custom Exceptions?
+          throw new RuntimeException("magic version " + magic + " is not supported")
```
cf. comment on corrupted indexes line 222.

Hangleton commented on core/src/main/scala/kafka/log/remote/RemoteLogIndexEntry.scala:
```diff
@@ -0,0 +1,229 @@
+            Some(RemoteLogIndexEntry(magic, crc, firstOffset, lastOffset, firstTimestamp, lastTimestamp, dataLength, rdi))
+          } catch {
+            case e: BufferUnderflowException =>
+              // TODO decide how to deal with incomplete records
```
Would it be possible at all to deal with an incomplete record?

ying-zheng committed:
incoporate review comments

ying-zheng committed:
Merge pull request #26 from ying-zheng/index_lookup

offset and timestamp lookup for remote data

ying-zheng committed:
Fix resource leak in HDFSRemoteStorageManager

ying-zheng committed:
Add in-memory read cache for HDFS (#28)

Add in-memory read cache for HDFS

satishd committed:
Fixed build issues with multiple scala versions

ying-zheng committed:
Fix the offset returned for NEXT_LOCAL_TIMESTAMP

satishd committed:
Added new SPI/API interfaces
More API changes and refactoring.

satishd committed:
Added refactored classes for RSM and dependent classes.

Hangleton committed:
Add an option to deactivate actual delete in the LocalRemoteStorageManager. Eases testing by allowing simulation of non-strongly consistent storage systems which do not guarantee visibility of a successful delete for subsequent read or list operations.

satishd committed:
Initial version of RemoteLogMetadataManager with topic based storage.
More refactoring of other classes with updated APIs.

Hanglet0n committed:
Tiered storage tests (#31)

* Add local file-system based implementation of RemoteStorageManager.
* Add local remote storage verifier.

ivanyu committed:
Don't use Optional in RemoteStorageManager.fetchLogSegmentData

harshach committed:
Merge pull request #32 from ivanyu/rsm-no-optional

Don't use Optional in RemoteStorageManager.fetchLogSegmentData

ying-zheng committed:
Rewrite HDFS RSM using the new interface

harshach committed:
Merge pull request #34 from Hangleton/tiered-storage-tests

Add an option to deactivate actual delete in the LocalRemoteStorageMa…

satishd committed:
Added initial fetch implementation for RLM and more.
 - Added initial fetch implementation for RLM and more fixes.
 - Added RLMM topic as internal topic and respective changes.
 - Added remote index cache.
 - More fixes, refactoring and cleanup.
 - Fixed a few spotBug and checkstyle issues.

Added RLMM topic as an internal topic with respective changes. This is done to run the default RemoteLogMetadataManager storage easily. We may want to change this later as an independent topic instead of an internal topic.

satishd committed:
Fixed RLMMWithTopicStorage to return segmentsids which contain the given offset.
 - Added test for edge cases.

ying-zheng committed:
Merge branch 'tiered-storage' into hdfs_rsm_new_api

ying-zheng committed:
Merge pull request #33 from ying-zheng/hdfs_rsm_new_api

Rewrite HDFS RSM using the new interface

ying-zheng committed:
Modify RSM interface

satishd committed:
MINOR Refactored `deleteLogSegment` API to return void and always throw RemoteStorageException if the operation is failed.

harshach committed:
Merge pull request #39 from satishd/minor-api-fix

MINOR Refactored `deleteLogSegment` API to return void and always throw RemoteStorageException if the operation is failed.

satishd committed:
Fixed a few boundary issues in RLMMWithTopicStorage

satishd committed:
Fixed a few failing tests.

harshach committed:
Merge pull request #40 from satishd/rlmm-changes

Fixed a few boundary issues in RLMMWithTopicStorage

satishd committed:
Fixed local and effective logStartOffset updation when segments are deleted locally but not in the remote tiered storage.

satishd committed:
Added option in ConsoleConsumer.DefaultFormatter to print offsets and other checkstyle and spotbug fixes.

ying-zheng committed:
Merge pull request #42 from satishd/log-start-offset-fix

Fixed local and effective logStartOffset updation when segments are deleted locally but not in the remote tiered storage.

satishd committed:
Added RLMM to consume only for leader/follower partitions that broker is hosted.

ying-zheng committed:
Fix EOF of remote input stream

satishd committed:
Merge pull request #44 from ying-zheng/eof

Fix EOF handling in the InputStream returned by HDFS RSM

satishd committed:
Removed FileChannel#transferFrom call and using Files.copy while fetching indexes from RSM to build index cache.

FileChannel#transferFrom call may be efficient as it goes through a fast path of transferring directly from the source channel into the filesystem cache. But if it goes through non-fast path then it expects the inputStream to always have available bytes. This is an unnecessary restriction on RemoteStorageManager to always return InputStream to have available bytes till the end.

satishd committed:
Added config for retention policy for remote log metadata topic and other minor fixes.

satishd committed:
Added more logging in RLMMWithTopicStorage.

satishd committed:
Added serdes for the messages stored in remote log metadata topic with internal protocol schema.

 - Added serdes for the messages stored in remote log metadata topic with internal protocol schema
 - Refactored RLMMWithTopicStorage to be more modular.

satishd committed:
Added RemoteLogSegmentMetadata ser/des for storing the snapshot of the recceived messages from remote log metadata topic.

 - Added tests for RLMM on leader and follower events from remote log metadata topic.

satishd committed:
RemoteLogMetadataManager#putRemoteLogSegmentData to take RemoteLogSegmentMetadata arg only.

- RemoteLogMetadataManager#putRemoteLogSegmentData to take RemoteLogSegmentMetadata arg only.

ying-zheng committed:
Fix several bugs related with remote log segment reading

satishd committed:
Optimized RLM fetch to read only once from RSM to get the desired batch of messages.

satishd committed:
Minimum one message handling is hanlded in RemoteLogManager.read when fetch max bytes is less than a single batch of message.

ying-zheng committed:
Merge pull request #49 from satishd/rlm-fetch-optimizations

Optimized RLM fetch to read only once from RSM to get the desired batch of messages.

ying-zheng committed:
Fix build error

ying-zheng committed:
Merge pull request #50 from ying-zheng/fix_build_error

Fix build error

Hanglet0n committed:
Enhanced Tiered storage test utils and added more tests.

- Add an option to deactivate actual delete in the LocalRemoteStorageManager. Eases testing by allowing simulation of non-strongly consistent storage systems which do not guarantee visibility of a successful delete for subsequent read or list operations.
- Add visitor to traverse locally emulated remote storage to provide support for test assertions.
- Add local remote storage listener to allow tests to be notified of modifications in the storage; - Add snaphost support for the local remote storage.
- Add waiter on local remote storage to allow tests to formulate expectations on asynchronously populated remote storage.
- Set readOffset's initial value to -1 to handle the edge case of a single record for the first segment of a topic-partition.
- Use the index size instead of log segment size to enable the creation of a single-record log segment.
- Added another segment to be offloaded in base integration test. Couple of fixes in test assertions.
- Add a test case for a segment with multiple records.

satishd committed:
Added more boundary test cases for RLMM RLMMWithTopicStorageTest.

ying-zheng committed:
Handled a boundary scenario RemoteLogManager.read for the target offset.

Handled a boundary scenario RemoteLogManager.read for the target offset.

ying-zheng committed:
Added timestamp lookup in remote storage.

Added timestamp lookup in remote storage.

satishd committed:
Fixed leadership changes of partitions in RLMM as updates only but not the complete set of partitions the broker hosted. (#54)

* Fixed leadership changes of partitions in RLMM as updates only but not the complete set of partitions the broker hosted.

ivanyu committed:
Add S3RemoteStorageManager implementation.

- Add S3RemoteStorageManager implementation.

satishd committed:
Fixed RemoteIndexCacheTest tests.

satishd committed:
Removed unused classes of earlier remtoe log index implementation.

satishd committed:
Minor cleanup in Tiered storage tests/utils and other checkstyle issues.

satishd committed:
Merge remote-tracking branch 'apache-repo/trunk' into merge-2.4.1-t-s

satishd committed:
Fixed build failures with the uptake of trunk branch.

satishd committed:
Minor fixes in RLMMWithTopicStorage to call configure only once.
 - Minor fixes in RLMMWithTopicStorage to call configure only once
 - Made requried fields to be volatile to avoid any stale reads being updated/read in different threads.

satishd committed:
Merge pull request #56 from satishd/merge-trunk-on-tiered-storage

Merge trunk branch on to tiered storage.

satishd committed:
Fixed AbstractIndex to update _file for its file operations

Hanglet0n committed:
Added Consume records and other tests, and refactoring of tiered storage test utils.

 - Added consume records and other tests, and refactoring of tiered storage test utils.  
 - Added broker id to the configuration properties of RSM. 
 - Added test for disaster recovery with deletion of broker local storage on leader.
(#52 )

Reviewed-by: satishd@apache.org

Hanglet0n committed:
Support producing to a log segment already contained records in produce-action in tiered storage tests

- Support producing to a log segment already contained records in produce-action.
- Generate basic test reports with one description per test action.
- Add utility to dump files from local tiered storage.
- Fix condition on local broker storage (filter out inactive and non-assigned brokers).

Reviewed-by: satishd@apache.org

Hanglet0n committed:
Add the broker ID which offloaded a given fileset to metadata and filename to provide it in local tiered storage dumps.

Add the broker ID which offloaded a given fileset to metadata and filename to provide it in local tiered storage dumps.

Reviewd-by: satishd@apache.org

Hanglet0n committed:
Support batch size > 1 in tiered storage integration tests. (#60)

* Support batch size > 1 in tiered storage integration tests.
* Enforce earliest offset in log directory equality rather than lower bound.
* Add all records and offsets to the tiered storage content dump.
* Explicit the records to be found in remote log segments in integration tests.

Reviewed-by: Satish Duggana<satishd@apache.org>

Hangleton committed:
Add test case for consumption from a follower.

ying-zheng committed:
Merge remote-tracking branch 'apache/trunk' into tiered-storage

satishd committed:
Send aborted transactions when fetch isolation is TxnCommitted for fetch requests and a few other changes.

  - Send aborted transactions when fetch isolation is TxnCommitted for fetch requests.
  - Minor cleanups.
  - Added producerIdsnapshot and txn indexes
  - Added follower catchup for remote indexes
  - Few changes in RSM and RLMM APIs and respective refactoring.
  - Added listener name config for RLMM.
  - Fixed a few minor issues.

ying-zheng committed:
Fix deadlock in remote metadata ConsumerTask.

Fix deadlock in remote metadata ConsumerTask.

satishd committed:
Refactored RemoteLogSegmentMetadata, RSM.copyLogSegment to take RemoteLogSegmentMetadata instead of RemoteLogSegmentId and a few other changes.

 - Refactored RemoteLogSegmentMetadata usages.
 - Updated RSM#copyLogSegment to take RemoteLogSegmentMetadata instead of RemoteLogSegmntId and replaced their usages.
 - Removed passing broker endpoints as part of onServerStarted callback as thosre properties are passed with config() method.

satishd committed:
RemoteLogMetadataManager takes leaderEpoch for some of the APIs to return respective offset or segment

  - RemoteLogMetadataManager takes leaderEpoch for some of the APIs to return respective offset or segment
  - Minor cleanups in the code.
  - Added states in RemoreLogSegmentMetadata.

tarunrathor-pro commented:
Curious whether this is still actively worked upon?

satishd committed:
Added leader epoch state to log segment data which is passed to RemoteStorageManager.

 - Added leader epoch state to log segment data which is passed to RemoteStorageManager.
 - Minor clean up with javadocs.

satishd committed:
Changed remote log metadata topic retention from minutes to milli seconds and other minor changes.

 - Changed remote log metadata topic retention from minutes to milli seconds.
 - Updated few javadocs for RSM and RLMM.
 - Fixed KafkaConfigTest.
 - RLMM#deleteSegment API is changed to be consistent wiht other RLMM APIs.

satishd committed:
Merged and resolved conflicts from upstream apache/trunk.

mgarmes commented on core/src/main/scala/kafka/log/remote/RemoteIndexCache.scala:
```diff
@@ -0,0 +1,218 @@
+import org.apache.kafka.common.utils.{KafkaThread, Utils}
+
+object RemoteIndexCache {
+  val DirName = "remote-log-index-cache"
```
Hello @satishd, I am testing the PR locally and I faced a problem during the kafka restart (  org.apache.kafka.common.KafkaException: Found directory /tmp/kafka-logs/remote-log-index-cache, 'remote-log-index-cache' is not in the form of topic-partition or topic-partition.uniqueId-delete (if marked for deletion).
Kafka's log directories (and children) should only contain Kafka topic data.)
bests.

satishd replied:
This is resolved with the latest commit.

mgarmes commented on remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManagerConfig.java:
```diff
@@ -0,0 +1,146 @@
+ * A configuration for {@link S3RemoteStorageManager}.
+ */
+public class S3RemoteStorageManagerConfig extends AbstractConfig {
+    public static final String S3_BUCKET_NAME_CONFIG = "s3.bucket.name";
```
Hello @satishd  I guess you miss the prefix  REMOTE_STORAGE_MANAGER_CONFIG_PREFIX.

ivanyu replied:
Maybe it's worth doing `config.originalsWithPrefix(REMOTE_STORAGE_MANAGER_CONFIG_PREFIX)` instead of adding the prefix manually to all configurations. I'll look into this.

ivanyu replied:
Fixed now

mgarmes commented on build.gradle:
```diff
@@ -1932,6 +1954,82 @@ project(':connect:basic-auth-extension') {
+  }
+
+  dependencies {
+    compile group: 'com.amazonaws', name: 'aws-java-sdk-s3', version: '1.11.726'
```
hello @satishd  I think you need to update 'compile' by 'localDeps' in order to generate "dependant-libs".

satishd committed:
Fixed to filter RemoteIndexCache.DirName in log dirs to traverse for partition dirs.

  - Fixed to filter RemoteIndexCache.DirName in log dirs to traverse for partition dirs. This check was there earlier but trunk merge changes removed it.
  - Minor javadoc updates.

satishd committed:
Initialized RLMM with respective props and added a test to make sure topic props are initialized with the same.

 - Fixed initializing RLMM with respective props.
 - Added a test to make sure topic props are initialized with the same.

satishd committed:
Passing endpoints to RemoteLogManager and then initializing RSM and RLMM with their configs.

 - Passing endpoints to RemoteLogManager and then initializing RSM and RLMM with their configs.
 - Removed onServerStarted in RLM and RLMM as it is not requried with this change. RSM.configure and RLMM.configure are invoked when the endpoint is available and before server starts accepting the requests.
 - Added a test to make sure initialized is invoked for the requried methods in RLMMWithTopicStorage.

satishd committed:
Merge remote-tracking branch 'apache-repo/trunk' into merge-trunk

ying-zheng committed:
Tiered storage new metrics

satishd committed:
Merge remote-tracking branch 'apache-repo/trunk' into merge-trunk

satishd committed:
Merge branch 'apache-trunk' into tiered-storage.

ivanyu committed:
Add prefix to S3RemoteStorageManagerConfig

The configurations were missing `remote.log.storage.` prefix.

ivanyu committed:
Improve S3RemoteStorageManager.close()

satishd committed:
Merge branch 'apache-trunk' into tiered storage.

kamalcph committed:
Fix OffloadAndConsumeFromFollowerTest tiered storage test.

- Disabled the CanFetchFromTieredStorageAfterRecoveryOfLocalSegmentsTest
  as it's not handled in the code.
- Verified that the records are served from the follower by enabling the
  request loggers. There is also a  check in the test to assert that the
  records are being fetched from the tiered storage in the follower node.
- Reduced the number of remote log metadata topic partitions to 5.

satishd committed:
Merge branch 'apache-trunk' into merge-trunk

kamalcph committed:
MINOR: Fix the reported checkstyle and Spotbugs issues to run the build.

Command: `./gradlew clean build -x signArchives -x test`

kamalcph committed:
Add support to copy txn, leader_epoch checkpoint & snapshot files in LocalTieredStorage.

- Fixed the truncation of leader epoch checkpoint after uploading the segment.
- Minor code cleanups.
- added test to verify the upload of index & snapshot files.

kamalcph committed:
Added message formatter to view the contents of remote metadata log.

- The remote log metadata content can be viewed using kafka-console-consumer (or) kafka-dump-log script.
- sh kafka-dump-log.sh --files <segment_file> --remote-metadata-log-decoder -print-data-log
- sh kafka-console-consumer.sh --bootstrap-server <listeners> --topic __remote_log_segment_metadata --from-beginning --formatter kafka.log.remote.RemoteMetadataLog\$RMLMessageFormatter

kamalcph committed:
Fix the deletion of log segment before uploading to remote tier.

kamalcph committed:
Follower fetch protocol v1 for tiered storage.

kamalcph committed:
Add list offset request V0 support for earliest local timestamp.

- added unit tests.

satishd commented:
We decided to have several PRs for tiered storage with the respective tasks for KAFKA-7739. This it to make the review process smooth based on the request from the reviewers, and it has also become difficult to add new changes into this branch and rebase with the latest trunk.

satishd commented:
We decided to have several PRs for tiered storage with the respective tasks for KAFKA-7739. This is to make the review process smooth based on the request from the reviewers, and it has also become difficult to add new changes from our internal branches and rebase with the latest trunk.
